training-unet-instance-segmentation/picsellia/utils.py

```diff
@@ -0,0 +1,286 @@
+
+
+class Dataset:
+    def __init__(
+        self,
+        images_dir,
+        masks_dir,
+        classes=None,
+        augmentation=None,
+        preprocessing=None,
+    ):
+        self.ids = os.listdir(images_dir)
+        self.images_filenames = [
+            os.path.join(images_dir, image_id) for image_id in self.ids
+        ]
+        self.masks_filenames = [
+            os.path.join(
+                masks_dir,
+                image_id.split(".")[0]
+                + "."
+                + get_mask_file_extension(os.listdir(masks_dir)[0]),
+            )
+            for image_id in self.ids
+        ]
+        self.class_values = [classes.index(cls.lower()) for cls in classes]
+        self.augmentation = augmentation
+        self.preprocessing = preprocessing
+
+    def __getitem__(self, i) -> tuple[np.ndarray, np.ndarray]:
+        image = cv2.imread(self.images_filenames[i])
+        image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
+        mask = cv2.imread(self.masks_filenames[i], 0)
+
+        masks = [(mask == v) for v in self.class_values]
+        mask = np.stack(masks, axis=-1).astype("float")
+
+        if mask.shape[-1] != 1:
+            background = 1 - mask.sum(axis=-1, keepdims=True)
+            mask = np.concatenate((mask, background), axis=-1)
+
+        if self.augmentation:
+            sample = self.augmentation(image=image, mask=mask)
+            image, mask = sample["image"], sample["mask"]
+
+        if self.preprocessing:
+            sample = self.preprocessing(image=image, mask=mask)
+            image, mask = sample["image"], sample["mask"]
+
+        return image, mask
+
+    def __len__(self) -> int:
+        return len(self.ids)
+
+
+def get_mask_file_extension(mask_file_path: str) -> str:
+    return mask_file_path.split(".")[1]
+
+
+def extract_classes_from_mask(mask: np.ndarray, class_values: list) -> np.ndarray:
+    masks = [(mask == v) for v in class_values]
+    mask = np.stack(masks, axis=-1).astype("float")
+    return mask
+
+
+class Dataloader(keras.utils.Sequence):
+    def __init__(self, dataset: Dataset, batch_size=1, shuffle=False):
+        self.dataset = dataset
+        self.batch_size = batch_size
+        self.shuffle = shuffle
+        self.indexes = np.arange(len(dataset))
+
+        self.on_epoch_end()
+
+    def __getitem__(self, i) -> list:
+        start = i * self.batch_size
+        stop = (i + 1) * self.batch_size
+        data = []
+        for j in range(start, stop):
+            data.append(self.dataset[j])
+
+        batch = [np.stack(samples, axis=0) for samples in zip(*data)]
+
+        return batch
+
+    def __len__(self) -> int:
+        number_of_batches_per_epoch = len(self.indexes) // self.batch_size
+        return number_of_batches_per_epoch
+
+    def on_epoch_end(self):
+        if self.shuffle:
+            self.indexes = np.random.permutation(self.indexes)
+
+
+def get_training_augmentation() -> A.Compose:
+    train_transform = [
```

Comment:
I think we need to think about a strategy for data augmentation in the future, I guess defining a hard-coded augmentation list might not suit the user's use-case. Maybe something like an `augmentation_config.json`, something the user can change when strating the container instead of rebuilding the image. idk what do you think @PN-picsell?

Reply:
I guess we can mount a volume between the container and the host where the config files are stored. We can also provide a argument to specify with configuration file use in case there's several data_augmentation files. Something like:

```bash
docker run -v /training-unet-instance-segmentation/data_augmentation_files/:/data_augmentation_files picsellia/training-unet:latest
--data_augmentation_file_path /data_augmentation_files/config.json 
```

Maybe a strategy that looks like this:
- If there's no config file, or the config file path is wrong, use a default configuration
- Otherwise, load the config

We just need to think the best way to store those augmentations. In a `.py` file ? a `.json` file ?
I think this volume argument should stay optional and only be used when the user needs to experiment several data augmentations without rebuidling the docker image.